woocommerce/google-listings-and-ads
# Add extra product attributes

### Changes proposed in this Pull Request:



Part of #16 .

This PR adds `condition`, `is_bundle`, and `multipack` product attributes. These attributes are available for simple products and variations (`is_bundle` can be only set on the parent product if it's a variable).

Integration with the [WooCommerce Product Bundles](https://woocommerce.com/products/product-bundles/) extension is also added. This means that all of the attributes available to `simple` products are also applicable to `bundle` products. The `is_bundle` attribute is an exception because it will be automatically set to `true` for `bundle` products and no input will be displayed for it on a `bundle` product.

The product pricing is a bit complex when it comes to bundles. Bundles can have either a constant price (set in the `General` section) or a dynamic price based on the items they contain. Bundle items can also be optional or may contain product variations causing them to require user input before purchase. 

For now, we only support simple bundles that either have a constant price or require no user input on purchase. This means that we will not sync product bundles that contain variable products or optional items.

In addition to the above, this PR contains the following tweaks:

- Move the `custom-inputs` JS into separate modules
- Use a `BooleanSelect` input type instead of checkbox for boolean attributes (e.g. `Adult`, and `IsBundle`). This select only contains `Yes` and `No` values along with a `Default`. The main reason I made this change was because there was no way to set a value to be `Default` using a checkbox (i.e. it should be either true or false). And I think having a "Default" value would be useful when we implement category-level or system-level default values for attributes.
- For the same reason as above, I also modified the `Number` input to be a `type="text"` input instead of a `type="number"` and enforced the value to be an integer using JS. We now have an `Integer` and `Decimal` input.
- Added a new filter called `gla_attribute_hidden_product_types_{$attribute_id}` that can be used to hide a certain attribute for the given product types. This was needed for the `IsBundle` attribute which should be hidden on the `bundle` type product because it's always `true`.
- Added new methods to `Form` and `AttributeForm` classes to better manage their children and allow replacing the input for an attribute.
- Add the `gla_product_attribute_value_price` and `gla_product_attribute_value_sale_price` filters to allow third parties to modify the price of a product before submitting it to Google
- Add the `gla_get_sync_ready_products` filter to allow modifying the list of products that are ready to be synced. This is used in the Product Bundles integration to skip bundle products that require user input before purchase.

### Detailed test instructions:

#### Test attributes
1. Create a new product or edit an existing one
2. If it's a simple product: change the values of `condition`, `is_bundle`, and `multipack` in the "Google Listings and Ads" tab in the product data section
3. If it's a variable product: change the values of `condition`, and `multipack` for each variation in the "Google Listings and Ads" section under each variation. You can also set the Adult attribute for the variable product (parent) which will apply to all variations.
4. Save the product and make sure it's synced and variables are correctly updated in Google Merchant Center

#### Test bundle integration
1. Install and activate [WooCommerce Product Bundles](https://woocommerce.com/products/product-bundles/)
2. Create a bundle product and add only simple product items to it
3. Set a constant price for it via the General tab
4. Set some of the attributes in Google Listings and Ads section
5. Save the product and make sure it's synced and variables are correctly updated in Google Merchant Center

#### Test bundle integration with individual prices
1. Create a bundle product and add only simple product items to it
2. Set individual prices for each item in bundle by checking `Priced Individually` and adding a discount.
3. Save the product and make sure it's synced.
4. The price of the bundle should be reflected correctly (should be the same as in the shop) in Merchant Center

#### Test bundle integration with variable products
1. Create a bundle product and add simple and variable items to it
2. Save the product and make sure the product has **NOT** synced to Merchant Center
3. Create another bundle product and add simple items but make them optional
4. Save the product and make sure the product has **NOT** synced to Merchant Center

### Changelog Note:

> Tweak - Add extra product attributes + WooCommerce Product Bundles integration
